connectors/bitbucket-server/src/main/java/com/vmware/connectors/bitbucket/server/BitbucketServerController.java

```diff
@@ -0,0 +1,483 @@
+                            .buildAndExpand(
+                                    ImmutableMap.of(
+                                            PROJECT_KEY, pullRequest.getProjectKey(),
+                                            REPOSITORY_SLUG, pullRequest.getRepositorySlug(),
+                                            PULL_REQUEST_ID, pullRequest.getPullRequestId(),
+                                            ACTION, bitBucketServerAction.getAction()
+                                    )
+                            )
+                            .toUri();
+
+                    // This header is added for skipping the CSRF check. Otherwise, Atlassian denies the pull request ACTION.
+                    headers.add(ATLASSIAN_TOKEN, "no-check");
+
+                    final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                            uri,
+                            HttpMethod.POST,
+                            new HttpEntity<>(headers),
+                            String.class);
+
+                    return Async.toSingle(response);
+                });
+
+    }
+
+    private HttpHeaders buildHeaders(final String authHeader) {
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(AUTHORIZATION, authHeader);
+        return headers;
+    }
+
+    private Single<String> getVersion(final HttpHeaders headers,
+                              final String baseUrl,
+                              final BitbucketServerPullRequest pullRequest) {
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = getPullRequestInfo(headers, pullRequest, baseUrl);
+        return Async.toSingle(response)
+                .map(entity -> entity.getBody())
+                .map(jsonDocument -> Integer.toString(jsonDocument.read("$.version")));
+
+    }
+
+    private Observable<Card> getCardsForBitbucketServerPR(final HttpHeaders headers,
+                                                          final BitbucketServerPullRequest pullRequest,
+                                                          final String baseUrl,
+                                                          final String routingPrefix) {
+        logger.debug("Requesting pull request info from bitbucket server base url: {} and pull request info: {}", baseUrl, pullRequest);
+
+        final Single<ResponseEntity<JsonDocument>> bitBucketServerResponse = Async.toSingle(getPullRequestInfo(headers, pullRequest, baseUrl));
+        final Single<List<String>> comments = getComments(baseUrl, headers, pullRequest);
+
+        return Observable.zip(bitBucketServerResponse.toObservable(), comments.toObservable(), Pair::of)
+                .onErrorResumeNext(ObservableUtil::skip404)
+                .map(pair -> convertResponseIntoCard(pair.getLeft(), pullRequest, routingPrefix, pair.getRight()));
+    }
+
+    private ListenableFuture<ResponseEntity<JsonDocument>> getPullRequestInfo(final HttpHeaders headers,
```

Comment:
The result of this is always converted to a `Single` by the caller. May as well just return a `Single`.

Reply:
I have pull your changes into my branch.